### PR TITLE
Port ImmediateRenderer from scenegraph

### PR DESCRIPTION
**Pull request by grokys**
This PR replaces the current `Renderer` with `ImmediateRenderer` from `scenegraph` and ports across the related infrastructure. The two renderers mostly behave the same in that they render the visual tree without using an intermediate scenegraph, however `ImmediateRenderer` in this PR uses the same interfaces used by `DeferredRenderer` in `scenegraph`.

Changes:
- `Renderer` renamed `ImmediateRenderer` and `IRenderer` interface updated
- `RendererMixin` removed and logic moved into `ImmediateRenderer`
- Visual brushes are now rendered by the renderer itself rather than in the rendering backends. This is needed by the deferred renderer. To support this, two new interfaces are introduced: `IVisualBrushRenderer` and `IVisualBrushInitialize`
- `RenderTargetBitmap`s now accept a DPI on creation
- Parent control is now passed on visual tree attach/detach (not strictly needed in this PR but was brought over with updated unit tests)
- FPS rendering moved to `RendererBase` as it is shared by `DeferredRenderer` on `scenegraph` branch

This hopefully gets us really close to being able to merge `scenegraph` cleanly.

### CODE REVIEW

File: src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+            catch (RenderTargetCorruptedException ex)
+            {
+                Logging.Logger.Information("Renderer", this, "Render target was corrupted. Exception: {0}", ex);
+                _renderTarget.Dispose();
+                _renderTarget = null;
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Resized(Size size)
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs]: Call to `PlatformImpl.Invalidate(new Rect(clientSize));` in `TopLevel::HandleResized` was replaced by `renderer.Resized`, but Invalidate isn't called here. Is that intended?

File: src/Gtk/Avalonia.Gtk/GtkPlatform.cs [hunk 2]
```diff
@@ -115,7 +115,7 @@ public IPopupImpl CreatePopup()
 
         public IRenderer CreateRenderer(IRenderRoot root, IRenderLoop renderLoop)
```
Comment (kekekeks) [inline src/Gtk/Avalonia.Gtk/GtkPlatform.cs]: I'm not sure that `CreateRenderer` should live in platform code. I'd rather move it to `ITopLevelImpl` instead.

File: src/Skia/Avalonia.Skia/PlatformRenderInterface.cs [hunk 3]
```diff
@@ -67,12 +66,11 @@ public IBitmapImpl LoadBitmap(PixelFormat format, IntPtr data, int width, int he
...
-        public IRenderer CreateRenderer(IRenderRoot root, IRenderLoop renderLoop)
-        {
-            return new Renderer(root, renderLoop);
-        }
-
-        public IRenderTargetBitmapImpl CreateRenderTargetBitmap(int width, int height)
+        public IRenderTargetBitmapImpl CreateRenderTargetBitmap(
+            int width,
+            int height,
+            double dpiX,
```
Comment (kekekeks) [inline src/Skia/Avalonia.Skia/PlatformRenderInterface.cs]: `dpiX` and `dpiY` aren't passed anywhere

File: src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs (hunk 1 above)
Comment (grokys) [inline src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs]: From what I could tell, windowing systems automatically invalidate the window on a resize so it should be unnecessary.

File: src/Gtk/Avalonia.Gtk/GtkPlatform.cs (hunk 2 above)
Comment (grokys) [inline src/Gtk/Avalonia.Gtk/GtkPlatform.cs]: Yeah, that's do-able I think. I will move it there.

File: src/Skia/Avalonia.Skia/PlatformRenderInterface.cs (hunk 3 above)
Comment (grokys) [inline src/Skia/Avalonia.Skia/PlatformRenderInterface.cs]: Yeah, they will be when `DeferredRenderer` is ported over though.

**Review by kekekeks — APPROVED** [review]

File: src/Gtk/Avalonia.Gtk/GtkPlatform.cs (hunk 2 above)
Comment (grokys) [inline src/Gtk/Avalonia.Gtk/GtkPlatform.cs]: Actually, there's a separate `IRenderFactory` interface that `CreateRenderer` is declared on - it just happens that the platform classes implement that as well. Is that OK, or would you still prefer it to be on the `ITopLevelImpl`?

**Comment by grokys** [comment]
Ok, I'm going to merge this as is, just to move things along. Let me know if you think `CreateRenderer` would be better placed in `ITopLevelImpl` instead of `IRendererFactory` and I can make that change.

**Pull request merged**